holoviews/core/dimension.py

```diff
@@ -486,7 +486,41 @@ def __init__(self, data, id=None, plot_id=None, **params):
         This class also has an id instance attribute, which
         may be set to associate some custom options with the object.
         """
+        from . import Dataset, DataError
         self.data = data
+
+        # Handle initializing the dataset property.
+        self._dataset = None
+        input_dataset = params.pop('dataset', None)
+        if type(self) is Dataset:
+            self._dataset = self
+        elif input_dataset is not None:
+            # Clone dimension info from input dataset with reference to new
+            # data. This way we keep the metadata for all of the dimensions.
+            try:
+                self._dataset = input_dataset.clone(data=self.data)
+            except DataError:
+                # Dataset not compatible with input data
+                pass
+        if self._dataset is None:
+            # Create a default Dataset to wrap input data
+            try:
+                kdims = list(params.get('kdims', []))
+                vdims = list(params.get('vdims', []))
+                dims = kdims + vdims
+                dataset = Dataset(
```

Comment:
Just wondering whether this will have performance implications. Generally I wouldn't expect it to be expensive to create a Dataset but there might be a bit of overhead. Since we already have some guarantees that the data is in a valid format maybe we can bypass all the initialization code and just provide it with the Interface, data and dimensions.